crates/typst-cli/src/download.rs

```diff
@@ -0,0 +1,204 @@
+use std::collections::VecDeque;
+use std::io::{self, ErrorKind, Read, Stdout, Write};
+use std::time::{Duration, Instant};
+
+use ureq::Response;
+
+/// Keep track of this many past download amounts
+const DOWNLOAD_TRACK_COUNT: usize = 5;
+
+/// A wrapper around [`ureq::Response`] that reads the response body in chunks
+/// over a websocket and displays statistics about its progress.
+///
+/// Downloads will _never_ fail due to statistics failing to print, print errors
+/// are silently ignored.
+///
+/// [`ureq::Response`]: https://docs.rs/ureq/2.7.1/ureq/struct.Response.html
+pub struct RemoteReader {
```

Comment:
I think there is not need to have this as a public type. The whole `RemoteReader::from_response(response).download()` thing can be encapsulated into a single public function `download_with_progress(response)`.

Reply:
Maybe `download_with_progress` could even directly take an URL and return a `Result<Vec<u8>, ureq::Error>`? So that the `ureq::get` part is not duplicated.

Reply:
Would we lose that? It would just pass on the `ureq::Error`.

Reply:
Oh..

> `Result<Vec<u8>, ureq::Error>`

Totally read over that... my bad.